addons/product/models/product.py

```diff
@@ -337,6 +337,14 @@ def _onchange_uom(self):
         if self.uom_id and self.uom_po_id and self.uom_id.category_id != self.uom_po_id.category_id:
             self.uom_po_id = self.uom_id
 
+    @api.onchange('default_code')
+    def _onchange_default_code(self):
+        if self.default_code and self.env['product.template'].search_count([('default_code', '=', self.default_code)]):
```

Comment:
Remove self (if self has already a id) from the domain here, to avoid matching itself.